File: obc_controller/can_worker.py
"""
CAN bus worker running in a QThread.

Handles:
  - Connecting / disconnecting the python-can bus
  - Periodic Message1 TX (500 ms) with optional ramp (soft-start)
  - Message2 RX with timeout alarm (5 s)
  - Safe-stop on disconnect (sends Control=1 for several cycles)
"""

from __future__ import annotations

import logging
import time
from typing import Optional

import can

from PySide6.QtCore import QThread, Signal, QMutex, QMutexLocker

from obc_controller.can_protocol import (
    MSG1_ID,
    MSG2_ID,
    CYCLE_MS,
    TIMEOUT_S,
    ChargerControl,
    Message1,
    Message2,
)

log = logging.getLogger(__name__)

SAFE_STOP_CYCLES = 5  # Send Control=1 this many times before disconnect


def _move_towards(current: float, target: float, max_step: float) -> float:
    """Move *current* towards *target* by at most *max_step*."""
    diff = target - current
    if abs(diff) <= max_step:
        return target
    return current + (max_step if diff > 0 else -max_step)


class CANWorker(QThread):
    """Background thread that owns the python-can bus object."""

    # Signals emitted to the UI thread
    connected = Signal()
    disconnected = Signal()
    error = Signal(str)
    log_message = Signal(str)
    message2_received = Signal(object)  # Message2 dataclass
    timeout_alarm = Signal()
    tx_message = Signal(object)         # Message1 sent (for log)
    ramp_state = Signal(bool, float, float)  # active, ramped_v, ramped_a

    def __init__(self, parent=None):
        super().__init__(parent)
        self._bus: Optional[can.Bus] = None
        self._running = False
        self._mutex = QMutex()

        # TX target state (set from UI thread, protected by mutex)
        self._target_voltage: float = 0.0
        self._target_current: float = 0.0
        self._control: ChargerControl = ChargerControl.STOP_OUTPUTTING
        self._tx_enabled = False

        # Ramp config (set from UI thread, protected by mutex)
        self._ramp_enabled: bool = False
        self._ramp_rate_v: float = 5.0   # V/s
        self._ramp_rate_a: float = 0.5   # A/s
        self._ramp_reset_flag: bool = False

        # Connection parameters
        self._interface = "pcan"
        self._channel = "PCAN_USBBUS1"
        self._bitrate = 250000

    # ---- public setters (called from UI thread) --------------------------

    def set_connection_params(
        self, interface: str, channel: str, bitrate: int
    ) -> None:
        self._interface = interface
        self._channel = channel
        self._bitrate = bitrate

    def set_setpoints(self, voltage: float, current: float) -> None:
        with QMutexLocker(self._mutex):
            self._target_voltage = voltage
            self._target_current = current

    def set_control(self, ctrl: ChargerControl) -> None:
        with QMutexLocker(self._mutex):
            self._control = ctrl

    def set_ramp_config(
        self, enabled: bool, rate_v: float, rate_a: float
    ) -> None:
        with QMutexLocker(self._mutex):
            self._ramp_enabled = enabled
            self._ramp_rate_v = rate_v
            self._ramp_rate_a = rate_a

    def reset_ramp(self) -> None:
        """Request ramp state machine to restart from 0."""
        with QMutexLocker(self._mutex):
            self._ramp_reset_flag = True

    def enable_tx(self, enabled: bool) -> None:
        with QMutexLocker(self._mutex):
            self._tx_enabled = enabled

    def request_stop(self) -> None:
        self._running = False

    # ---- thread entry point ----------------------------------------------

    def run(self) -> None:  # noqa: C901
        # --- connect ---
        try:
            kwargs: dict = {
                "interface": self._interface,
                "channel": self._channel,
                "bitrate": self._bitrate,
            }
            if self._interface == "pcan":
                self.log_message.emit(
                    f"Connecting PCAN channel={self._channel} "
                    f"bitrate={self._bitrate} …"
                )
            self._bus = can.Bus(**kwargs)
            self.log_message.emit("CAN bus connected.")
            self.connected.emit()
        except Exception as exc:
            msg = f"CAN connect failed: {exc}"
            if self._interface == "pcan" and "bitrate" in str(exc).lower():
                msg += (
                    "\nNote: Bitrate is configured in PCAN driver / PCAN-View."
                )
            self.log_message.emit(msg)
            self.error.emit(msg)
            self.disconnected.emit()
            return

        self._running = True
        last_tx_time = 0.0
        last_rx_time = time.monotonic()
        alarm_active = False

        # Ramp internal state
        ramped_v = 0.0
        ramped_a = 0.0
        prev_control = ChargerControl.STOP_OUTPUTTING

        try:
            while self._running:
                now = time.monotonic()

                # ---- read UI state under lock ----------------------------
                with QMutexLocker(self._mutex):
                    tx_en = self._tx_enabled
                    ctrl = self._control
                    tgt_v = self._target_voltage
                    tgt_a = self._target_current
                    ramp_en = self._ramp_enabled
                    ramp_rv = self._ramp_rate_v
                    ramp_ra = self._ramp_rate_a
                    do_reset = self._ramp_reset_flag
                    self._ramp_reset_flag = False

                # ---- ramp reset triggers ---------------------------------
                if do_reset:
                    ramped_v = 0.0
                    ramped_a = 0.0

                # Transition from STOP to active mode -> reset ramp
                if (
                    ctrl != ChargerControl.STOP_OUTPUTTING
                    and prev_control == ChargerControl.STOP_OUTPUTTING
                ):
                    ramped_v = 0.0
                    ramped_a = 0.0
                prev_control = ctrl

                # ---- TX --------------------------------------------------
                if tx_en and (now - last_tx_time) >= (CYCLE_MS / 1000.0):
                    dt = now - last_tx_time if last_tx_time > 0 else CYCLE_MS / 1000.0

                    # Compute actual setpoints
                    if (
                        ctrl == ChargerControl.STOP_OUTPUTTING
                        or not ramp_en
                    ):
                        # No ramp: send target directly
                        send_v = tgt_v
                        send_a = tgt_a
                        ramp_active = False
                    else:
                        # Ramp towards target
                        step_v = ramp_rv * dt
                        step_a = ramp_ra * dt
                        ramped_v = _move_towards(ramped_v, tgt_v, step_v)
                        ramped_a = _move_towards(ramped_a, tgt_a, step_a)
                        # Round to 0.1 (CAN resolution)
                        send_v = round(ramped_v, 1)
                        send_a = round(ramped_a, 1)
                        ramp_active = (
                            send_v != round(tgt_v, 1)
                            or send_a != round(tgt_a, 1)
                        )

                    msg1 = Message1(
                        voltage_setpoint=send_v,
                        current_setpoint=send_a,
                        control=ctrl,
                    )
                    frame = can.Message(
                        arbitration_id=MSG1_ID,
                        data=msg1.encode(),
                        is_extended_id=True,
                    )
                    try:
                        self._bus.send(frame)
                        last_tx_time = now
                        self.tx_message.emit(msg1)
                        self.ramp_state.emit(ramp_active, send_v, send_a)
                    except can.CanError as exc:
                        self.log_message.emit(f"TX error: {exc}")

                # ---- RX --------------------------------------------------
                try:
                    frame = self._bus.recv(timeout=0.05)
                except can.CanError as exc:
                    self.log_message.emit(f"RX error: {exc}")
                    frame = None

                if frame is not None and frame.arbitration_id == MSG2_ID:
                    try:
                        msg2 = Message2.decode(frame.data)
                        self.message2_received.emit(msg2)
                        last_rx_time = now
                        if alarm_active:
                            alarm_active = False
                            self.log_message.emit(
                                "Message2 received — timeout cleared."
                            )
                    except Exception as exc:
                        self.log_message.emit(
                            f"Message2 decode error: {exc}"
                        )

                # ---- timeout check ---------------------------------------
                if tx_en and (now - last_rx_time) > TIMEOUT_S:
                    if not alarm_active:
                        alarm_active = True
                        self.timeout_alarm.emit()
                        self.log_message.emit(
                            "ALARM: No Message2 for > 5 s!"
                        )

        finally:
            self._safe_stop()
            self._close_bus()
            self.disconnected.emit()

    # ---- internal helpers ------------------------------------------------

    def _safe_stop(self) -> None:
        """Send Control=1 (stop) for several cycles before shutting down."""
        if self._bus is None:
            return
        self.log_message.emit("Safe-stop: sending Control=STOP …")
        msg1 = Message1(
            voltage_setpoint=0,
            current_setpoint=0,
            control=ChargerControl.STOP_OUTPUTTING,
        )
        frame = can.Message(
            arbitration_id=MSG1_ID,
            data=msg1.encode(),
            is_extended_id=True,
        )
        for _ in range(SAFE_STOP_CYCLES):
            try:
                self._bus.send(frame)
            except can.CanError:
                break
            time.sleep(CYCLE_MS / 1000.0)
        self.log_message.emit("Safe-stop complete.")

    def _close_bus(self) -> None:
        if self._bus is not None:
            try:
                self._bus.shutdown()
            except Exception:
                pass
            self._bus = None
            self.log_message.emit("CAN bus closed.")

File: cpp_app/can_worker.cpp
#include "can_worker.h"

#include <QThread>
#include <cstring>
#include <cmath>
#include <algorithm>
#include <chrono>

static double monoNow() {
    auto now = std::chrono::steady_clock::now();
    return std::chrono::duration<double>(now.time_since_epoch()).count();
}

static double moveTowards(double current, double target, double maxStep) {
    double diff = target - current;
    if (std::abs(diff) <= maxStep)
        return target;
    return current + (diff > 0 ? maxStep : -maxStep);
}

// === BaudrateSwitchWorker ===

static const uint8_t BAUD_FRAME1_DATA[] = {0x07, 0x01, 0x00, 0x00, 0x3D, 0x8A, 0x09, 0x00};
static const uint8_t BAUD_FRAME2_DATA[] = {0x07, 0x02, 0x0E, 0x00, 0x71, 0xB7, 0x0F, 0x00};

BaudrateSwitchWorker::BaudrateSwitchWorker(ICanIface* iface, QObject* parent)
    : QThread(parent), m_iface(iface) {}

void BaudrateSwitchWorker::requestStop() {
    m_running = false;
}

void BaudrateSwitchWorker::run() {
    if (!m_iface || !m_iface->isOpen()) {
        emit error("CAN interface not available for baudrate switch.");
        return;
    }

    int total = 1 + BAUD_FRAME2_COUNT;

    // Step 1: send frame #1
    CanFrame frame;
    frame.id = BAUD_SWITCH_ID;
    frame.isExtended = true;
    frame.dlc = 8;
    std::memcpy(frame.data, BAUD_FRAME1_DATA, 8);

    if (!m_iface->send(frame)) {
        emit error(QString("Baudrate switch frame #1 TX error: %1")
                       .arg(QString::fromStdString(m_iface->lastError())));
        return;
    }
    emit logMessage(QString("Baudrate switch: sent frame #1 (ID=0x%1)")
                        .arg(BAUD_SWITCH_ID, 8, 16, QChar('0')));
    emit progress(1, total);

    // Steps 2-8: send frame #2 seven times @ 500ms
    std::memcpy(frame.data, BAUD_FRAME2_DATA, 8);
    for (int i = 0; i < BAUD_FRAME2_COUNT; ++i) {
        if (!m_running) {
            emit logMessage("Baudrate switch aborted.");
            return;
        }
        QThread::msleep(static_cast<unsigned long>(BAUD_FRAME2_INTERVAL_S * 1000));
        if (!m_iface->send(frame)) {
            emit error(QString("Baudrate switch frame #2 [%1] TX error: %2")
                           .arg(i + 1)
                           .arg(QString::fromStdString(m_iface->lastError())));
            return;
        }
        emit logMessage(QString("Baudrate switch: sent frame #2 (%1/%2)")
                            .arg(i + 1).arg(BAUD_FRAME2_COUNT));
        emit progress(2 + i, total);
    }

    emit logMessage("Baudrate switch sequence completed.");
    emit finishedOk();
}

// === CANWorker ===

CANWorker::CANWorker(std::unique_ptr<ICanIface> iface, QObject* parent)
    : QThread(parent), m_iface(std::move(iface)) {}

void CANWorker::setSetpoints(double voltage, double current) {
    QMutexLocker lock(&m_mutex);
    m_targetVoltage = voltage;
    m_targetCurrent = current;
}

void CANWorker::setControl(ChargerControl ctrl) {
    QMutexLocker lock(&m_mutex);
    m_control = ctrl;
}

void CANWorker::setRampConfig(bool enabled, double rateV, double rateA) {
    QMutexLocker lock(&m_mutex);
    m_rampEnabled = enabled;
    m_rampRateV = rateV;
    m_rampRateA = rateA;
}

void CANWorker::resetRamp() {
    QMutexLocker lock(&m_mutex);
    m_rampResetFlag = true;
}

void CANWorker::enableTx(bool enabled) {
    QMutexLocker lock(&m_mutex);
    m_txEnabled = enabled;
}

void CANWorker::requestStop() {
    QMutexLocker lock(&m_mutex);
    m_running = false;
}

ICanIface* CANWorker::getInterface() const {
    return m_iface.get();
}

bool CANWorker::isBusConnected() const {
    return m_iface && m_iface->isOpen();
}

double CANWorker::calcRate(const std::deque<double>& times, double now, double window) const {
    double cutoff = now - window;
    int count = 0;
    for (double t : times) {
        if (t >= cutoff) ++count;
    }
    return static_cast<double>(count) / window;
}

void CANWorker::run() {
    if (!m_iface || !m_iface->isOpen()) {
        emit error("CAN interface not available or not open.");
        emit disconnected();
        return;
    }

    emit logMessage(QString("CAN bus connected via %1.")
                        .arg(QString::fromStdString(m_iface->backendName())));
    emit connected();

    {
        QMutexLocker lock(&m_mutex);
        m_running = true;
    }

    double lastTxTime = 0.0;
    double lastRxTime = monoNow();
    bool alarmActive = false;

    m_txTimes.clear();
    m_rxTimes.clear();
    m_prevStatusByte = -1;

    double rampedV = 0.0;
    double rampedA = 0.0;
    ChargerControl prevControl = ChargerControl::STOP_OUTPUTTING;

    while (true) {
        double now = monoNow();

        // Read UI state
        bool txEn;
        ChargerControl ctrl;
        double tgtV, tgtA;
        bool rampEn;
        double rampRV, rampRA;
        bool doReset;
        {
            QMutexLocker lock(&m_mutex);
            if (!m_running) break;
            txEn = m_txEnabled;
            ctrl = m_control;
            tgtV = m_targetVoltage;
            tgtA = m_targetCurrent;
            rampEn = m_rampEnabled;
            rampRV = m_rampRateV;
            rampRA = m_rampRateA;
            doReset = m_rampResetFlag;
            m_rampResetFlag = false;
        }

        // Ramp reset
        if (doReset) {
            rampedV = 0.0;
            rampedA = 0.0;
        }
        if (ctrl != ChargerControl::STOP_OUTPUTTING
            && prevControl == ChargerControl::STOP_OUTPUTTING) {
            rampedV = 0.0;
            rampedA = 0.0;
        }
        prevControl = ctrl;

        // TX
        if (txEn && (now - lastTxTime) >= (CYCLE_MS / 1000.0)) {
            double dt = (lastTxTime > 0) ? (now - lastTxTime) : (CYCLE_MS / 1000.0);

            double sendV, sendA;
            bool rampActive;
            if (ctrl == ChargerControl::STOP_OUTPUTTING || !rampEn) {
                sendV = tgtV;
                sendA = tgtA;
                rampActive = false;
            } else {
                double stepV = rampRV * dt;
                double stepA = rampRA * dt;
                rampedV = moveTowards(rampedV, tgtV, stepV);
                rampedA = moveTowards(rampedA, tgtA, stepA);
                sendV = std::round(rampedV * 10.0) / 10.0;
                sendA = std::round(rampedA * 10.0) / 10.0;
                rampActive = (sendV != std::round(tgtV * 10.0) / 10.0
                           || sendA != std::round(tgtA * 10.0) / 10.0);
            }

            Message1 msg1;
            msg1.voltage_setpoint = sendV;
            msg1.current_setpoint = sendA;
            msg1.control = ctrl;

            auto payload = msg1.encode();

            CanFrame txFrame;
            txFrame.id = MSG1_ID;
            txFrame.dlc = 8;
            txFrame.isExtended = true;
            std::memcpy(txFrame.data, payload.data(), 8);

            if (m_iface->send(txFrame)) {
                lastTxTime = now;
                m_txTimes.push_back(now);
                if (m_txTimes.size() > 20) m_txTimes.pop_front();
                emit txMessage(msg1);
                emit rampState(rampActive, sendV, sendA);
            } else {
                emit logMessage(QString("TX error: %1")
                                    .arg(QString::fromStdString(m_iface->lastError())));
            }

            double txRate = calcRate(m_txTimes, now);
            double rxRate = calcRate(m_rxTimes, now);
            double rxAge = now - lastRxTime;
            emit healthStats(txRate, rxRate, rxAge);
        }

        // RX (50ms timeout)
        CanFrame rxFrame;
        if (m_iface->recv(rxFrame, 50)) {
            if (rxFrame.id == MSG2_ID) {
                try {
                    Message2 msg2 = Message2::decode(rxFrame.data, rxFrame.dlc);
                    emit message2Received(msg2);
                    lastRxTime = now;
                    m_rxTimes.push_back(now);
                    if (m_rxTimes.size() > 20) m_rxTimes.pop_front();

                    if (alarmActive) {
                        alarmActive = false;
                        emit logMessage("Message2 received \u2014 timeout cleared.");
                    }

                    // Detect status bit changes
                    int newStatus = msg2.status.toByte();
                    if (m_prevStatusByte >= 0) {
                        int xorBits = newStatus ^ m_prevStatusByte;
                        for (int bit = 0; bit < 5; ++bit) {
                            if (xorBits & (1 << bit)) {
                                bool isFault = (newStatus & (1 << bit)) != 0;
                                emit statusBitChanged(bit, QString(statusBitName(bit)), isFault);
                            }
                        }
                    }
                    m_prevStatusByte = newStatus;
                } catch (const std::exception& e) {
                    emit logMessage(QString("Message2 decode error: %1").arg(e.what()));
                }
            }
        }

        // Timeout check
        if (txEn && (now - lastRxTime) > TIMEOUT_S) {
            if (!alarmActive) {
                alarmActive = true;
                emit timeoutAlarm();
                emit logMessage("ALARM: No Message2 for > 5 s!");
            }
        }
    }

    safeStop();

    // Close CAN bus
    if (m_iface) {
        m_iface->close();
        emit logMessage("CAN bus closed.");
    }

    emit disconnected();
}

void CANWorker::safeStop() {
    if (!m_iface || !m_iface->isOpen()) return;
    emit logMessage("Safe-stop: sending Control=STOP \u2026");

    Message1 msg1;
    msg1.voltage_setpoint = 0;
    msg1.current_setpoint = 0;
    msg1.control = ChargerControl::STOP_OUTPUTTING;
    auto payload = msg1.encode();

    CanFrame frame;
    frame.id = MSG1_ID;
    frame.dlc = 8;
    frame.isExtended = true;
    std::memcpy(frame.data, payload.data(), 8);

    for (int i = 0; i < SAFE_STOP_CYCLES; ++i) {
        if (!m_iface->send(frame))
            break;
        QThread::msleep(CYCLE_MS);
    }
    emit logMessage("Safe-stop complete.");
}
